llama-index-packs/llama-index-packs-secgpt/llama_index/packs/secgpt/spoke.py

```diff
@@ -0,0 +1,112 @@
+"""
+By integrating sandboxing, the spoke operator, and the spoke output parser with an LLM, memory, and app, we can build a standard spoke. We demonstrate the integration of these components below.
+"""
+
+from llama_index.llms.openai import OpenAI
+from llama_index.core.llms.llm import LLM
+
+from llama_index.core.memory.chat_memory_buffer import ChatMemoryBuffer
+
+from llama_index.core.tools import BaseTool
+from llama_index.core.settings import Settings
+
+from .tool_importer import create_message_spoke_tool, create_function_placeholder
+from .spoke_operator import SpokeOperator
+from .spoke_parser import SpokeOutputParser
+from .sandbox import set_mem_limit, drop_perms
+
+
+from typing import Sequence, List, Optional
+
+
+from llama_index.core.agent import ReActAgent
+
+from llama_index.core.base.llms.types import ChatMessage
+
+
+class Spoke:
+    def __init__(
+        self,
+        tools: Sequence[BaseTool],
+        collab_functions: Sequence[str],
+        llm: LLM = None,
+        memory: ChatMemoryBuffer = None,
+        verbose: bool = True,
+    ) -> None:
+        self.tools = tools
+        if self.tools:
+            self.tool_name = tools[0].metadata.name
+        else:
+            self.tool_name = ""
+
+        self.collab_functions = collab_functions
+        self.llm = llm or Settings.llm
+        self.memory = memory or ChatMemoryBuffer.from_defaults(
+            chat_history=[], llm=self.llm
+        )
+
+        # Set up spoke operator
+        self.spoke_operator = SpokeOperator(self.collab_functions)
+
+        # Create a placeholder for each collabortive functionality
+        func_placeholders = create_function_placeholder(self.collab_functions)
+
+        # Set the tool and collabortive functionality list
+        tool_functionality_list = (
+            self.tools + func_placeholders + [create_message_spoke_tool()]
+        )
+
+        # Set up the spoke output parser
+        self.spoke_output_parser = SpokeOutputParser(
+            functionality_list=self.collab_functions, spoke_operator=self.spoke_operator
+        )
+        # Set up the spoke agent
+        self.spoke_agent = ReActAgent.from_tools(
+            tools=tool_functionality_list,
+            llm=self.llm,
+            memory=self.memory,
+            output_parser=self.spoke_output_parser,
+            verbose=verbose,
+        )
+
+    def chat(
```

Comment:
I think since this is a light wrapper on our React class, I think you can support stream_chat and its async versions as well.

Reply:
I have included other chatting methods accordingly.